pkg/proxy/hybrid/proxy.go

```diff
@@ -146,7 +155,20 @@ func (p *HybridProxier) shouldEndpointsUseUserspace(endpoints *api.Endpoints) bo
 	return false
 }
 
+// switchService moves a service between the unidling and main proxies.
 func (p *HybridProxier) switchService(name types.NamespacedName) {
+	// We shouldn't call switchService more than once (per switch), but there
+	// are some logic bugs where this happens
+	// So, cache the real state and don't allow this to be called twice.
+	p.serviceSwitchedLock.Lock()
+	defer p.serviceSwitchedLock.Unlock()
+
+	switched, ok := p.serviceSwitched[name]
+	if ok && p.usingUserspace[name] == switched {
+		glog.V(6).Infof("hybrid proxy: ignoring duplicate switchService(%s/%s)", name.Namespace, name.Name)
+		return
+	}
+
```

Comment:
So we never remove services from serviceSwitched, right?  We probably need to do that otherwise we grow the map forever.  Can we do that from ServiceDelete()?